include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1235 @@
+MPI_Request i_all_to_all(RecvType* recv_buffer, const int recv_count,
+                         std::shared_ptr<const communicator> comm)
+{
+    MPI_Request req;
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Ialltoall(
+        detail::in_place<RecvType>(), recv_count, get_type<RecvType>(),
+        recv_buffer, recv_count, get_type<RecvType>(), comm->get(), &req));
+    return req;
+}
+
+
+/**
+ * Communicate data from all ranks to all other ranks (MPI_Alltoall).
+ * See MPI documentation for more details.
+ *
+ * @param send_buffer  the buffer to send
+ * @param send_count  the number of elements to send
+ * @param recv_buffer  the buffer to receive
+ * @param recv_count  the number of elements to receive
+ * @param comm  the communicator
+ */
+template <typename SendType, typename RecvType>
+void all_to_all(const SendType* send_buffer, const int send_count,
+                RecvType* recv_buffer, const int recv_count,
+                std::shared_ptr<const communicator> comm)
+{
+    GKO_ASSERT_NO_MPI_ERRORS(
+        MPI_Alltoall(send_buffer, send_count, get_type<SendType>(), recv_buffer,
+                     recv_count, get_type<RecvType>(), comm->get()));
+}
+
+
+/**
+ * Communicate data from all ranks to all other ranks (MPI_Alltoall).
+ * See MPI documentation for more details.
+ *
+ * @param send_buffer  the buffer to send
+ * @param send_count  the number of elements to send
+ * @param recv_buffer  the buffer to receive
+ * @param recv_count  the number of elements to receive
+ * @param comm  the communicator
+ *
+ * @return  the request handle for the call
+ */
+template <typename SendType, typename RecvType>
+MPI_Request i_all_to_all(const SendType* send_buffer, const int send_count,
+                         RecvType* recv_buffer, const int recv_count,
+                         std::shared_ptr<const communicator> comm)
+{
+    MPI_Request req;
+    GKO_ASSERT_NO_MPI_ERRORS(MPI_Ialltoall(
+        send_buffer, send_count, get_type<SendType>(), recv_buffer, recv_count,
+        get_type<RecvType>(), comm->get(), &req));
+    return req;
+}
+
+
+/**
+ * Communicate data from all ranks to all other ranks with
+ * offsets (MPI_Alltoallv). See MPI documentation for more details.
+ *
+ * @param send_buffer  the buffer to send
+ * @param send_count  the number of elements to send
+ * @param send_offsets  the offsets for the send buffer
+ * @param recv_buffer  the buffer to gather into
+ * @param recv_count  the number of elements to receive
+ * @param recv_offsets  the offsets for the recv buffer
+ * @param stride  the stride to be used in case of sending concatenated data
+ * @param comm  the communicator
+ */
+template <typename SendType, typename RecvType>
+void all_to_all_v(const SendType* send_buffer, const int* send_counts,
+                  const int* send_offsets, RecvType* recv_buffer,
+                  const int* recv_counts, const int* recv_offsets,
+                  const int stride, std::shared_ptr<const communicator> comm)
```

Comment:
stride is unused, what does it stand for?
```suggestion
                  const int stride, std::shared_ptr<const communicator> comm)
```